src/index.spec.ts

```diff
@@ -0,0 +1,168 @@
+describe('Demonstration of downstream testing', () => {
+  let app: App;
+  let handler: any;
+  let request: any;
+  const message = 'Oh Hai';
+
+  beforeEach(async () => {
+    const receiver = new ExpressReceiver({ signingSecret: 'SECRET' });
+    const RewiredApp = await importAppWithMockSlackClient();
+
+    app = new RewiredApp({ receiver, token: '' });
+
+    // Possibly this wrapper isn't the right one.
+    // See https://community.slack.com/archives/CHY642221/p1575577886047900 for details.
+    // This wrapper should take an event and return a promise with a response when it's
+    // event loop has completed
+    handler = serverless(receiver.app);
+
+    const event = createFakeMessageEvent(message);
+    request = createEventRequest(event);
+  });
+
+  it('correctly waits for async listeners', async () => {
+    let changed = false;
+
+    app.message(message, async ({ next }) => {
+      await delay(100);
+      changed = true;
+
+      next();
+    });
+
+    const response = await handler(request);
+
+    assert.equal(response.statusCode, 200);
+    assert.isTrue(changed); // Actual `true`
+  });
+
+  it('Calls async middleware in a nested, declared first, order', async () => {
+    let orderCount = 1;
+
+    const assertOrderMiddleware = (order: number) =>  async ({ next }: any) => {
+      await delay(100);
+      assert.equal(orderCount, order);
+      orderCount += 1;
+      next();
+    };
+
+    app.use(assertOrderMiddleware(1));
+
+    app.message(message, assertOrderMiddleware(2), assertOrderMiddleware(3));
+
+    // This middleware is never called, the middleware does a thing where if it
+    // detects a message (3 in this case) as "last" it will give it a noop
+    // instead of a real callback. Possibly interesting cases could be handled
+    // if everything is connected. I discovered this by trying to polyfill bolt sticking
+    // a handler here to possibly find when the event loop was done.
+    //
+    // A real use case would be having a message set a `state` in it's context and a
+    // handler here saving it to a db
+    app.use(assertOrderMiddleware(4));
+
+    await handler(request);
+
+    // This should be removable, without it none of the middleware is called
+    await delay(600);
+
+    assert.equal(orderCount, 5); // Actual 4
+  });
+
+  it('can catch thrown errors in downstream async listeners', async () => {
+    app.message('Hai', async ({next}) => {
+      const error =  new Error('Something terrible happened!');
+
+      // real "async" middleware wouldn't do this, but I would have expected this to work.
+      next(error);
+
+      // Nothing catches this up the stack, this is what async middleware is likely doing
+      throw error;
+    });
+
+    app.error(() => {
+      // This is never called, I do agree that middleware should handle it's own errors.
+      // Having a handler can be helpful for strange, unexpected ones.
+      // Also lazy (read me) developers!
```

Comment:
too cutesy to put as a comment in code. personal preference is to keep it to the facts, there's already a decent amount to digest here. IMO.